unexpected indent behaviour in multiline boolean expression in 4.x

<!--
    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).

    This template is for bug reports. If you are reporting a bug, please continue on. If you are here for another reason, please see below:

    1. To propose a new rule: http://eslint.org/docs/developer-guide/contributing/new-rules
    2. To request a change: http://eslint.org/docs/developer-guide/contributing/changes
    3. If you have any questions, please stop by our chatroom: https://gitter.im/eslint/eslint

    Note that leaving sections blank will make it difficult for us to troubleshoot and we may have to close the issue.
-->

**Tell us about your environment**

* **ESLint Version:** 4.2.0
* **Node Version:** 8.1.3
* **npm Version:** 5.0.3

**What parser (default, Babel-ESLint, etc.) are you using?**
babel-eslint

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

<!-- Paste your configuration below: -->
```json
{
    "parser": "babel-eslint",
    "env": {
        "es6": true,
        "node": true
    },
    "rules": {
        "indent": [
            "error",
            4,
            {
                "SwitchCase": 1,
                "MemberExpression": 1
            }
        ]
    }
}
```

</details>

**What did you do? Please include the actual source code causing the issue.**

<!-- Paste the source code below: -->
```js
const skip = (request, response) => {
    return first
        || (
            second
            && third
        );
};
```

**What did you expect to happen?**
I expected the code to pass the indent rule with given configuration.

**What actually happened? Please include the actual, raw output from ESLint.**
```
/…/src/index.js
  5:1  error  Expected indentation of 8 spaces but found 12  indent
  7:1  error  Expected indentation of 4 spaces but found 8   indent
```

However, the proposed fix looks rather odd
```js
const skip = (request, response) => {
    return first
        || (
        second // no indentation here?
            && third // inconsistent indentation of the first and second part of the condition
    ); // inconsistent indentation of the opening and closing brace
};
```

Is this intended behaviour? Thank you.

--- a/lib/rules/indent.js
+++ b/lib/rules/indent.js
@@ -896,27 +896,6 @@ module.exports = {
         }
 
         /**
-        * Adds indentation for the right-hand side of binary/logical expressions.
-        * @param {ASTNode} node A BinaryExpression or LogicalExpression node
-        * @returns {void}
-        */
-        function addBinaryOrLogicalExpressionIndent(node) {
-            const operator = sourceCode.getFirstTokenBetween(node.left, node.right, token => token.value === node.operator);
-
-            /*
-             * For backwards compatibility, don't check BinaryExpression indents, e.g.
-             * var foo = bar &&
-             *                   baz;
-             */
-
-            const tokenAfterOperator = sourceCode.getTokenAfter(operator);
-
-            offsets.ignoreToken(operator);
-            offsets.ignoreToken(tokenAfterOperator);
-            offsets.setDesiredOffsets([operator.range[1], node.range[1]], tokenAfterOperator, 1);
-        }
-
-        /**
         * Checks the indentation for nodes that are like function calls (`CallExpression` and `NewExpression`)
         * @param {ASTNode} node A CallExpression or NewExpression node
         * @returns {void}
@@ -1077,7 +1056,22 @@ module.exports = {
                 offsets.ignoreToken(sourceCode.getTokenAfter(operator));
             },
 
-            BinaryExpression: addBinaryOrLogicalExpressionIndent,
+            "BinaryExpression, LogicalExpression"(node) {
+                const operator = sourceCode.getFirstTokenBetween(node.left, node.right, token => token.value === node.operator);
+
+                /*
+                * For backwards compatibility, don't check BinaryExpression indents, e.g.
+                * var foo = bar &&
+                *                   baz;
+                */
+
+                const tokenAfterOperator = sourceCode.getTokenAfter(operator);
+
+                offsets.ignoreToken(operator);
+                offsets.ignoreToken(tokenAfterOperator);
+                offsets.setDesiredOffset(tokenAfterOperator, operator, 0);
+                offsets.setDesiredOffsets([tokenAfterOperator.range[1], node.range[1]], tokenAfterOperator, 1);
+            },
 
             BlockStatement: addBlockIndent,
 
@@ -1210,8 +1204,6 @@ module.exports = {
                     offsets.setDesiredOffsets([fromToken.range[0], node.range[1]], sourceCode.getFirstToken(node), 1);
                 }
             },
-
-            LogicalExpression: addBinaryOrLogicalExpressionIndent,
 
             "MemberExpression, JSXMemberExpression"(node) {
                 const firstNonObjectToken = sourceCode.getFirstTokenBetween(node.object, node.property, astUtils.isNotClosingParenToken);

--- a/tests/lib/rules/indent.js
+++ b/tests/lib/rules/indent.js
@@ -2753,6 +2753,22 @@ ruleTester.run("indent", rule, {
         },
         {
             code: unIndent`
+                foo
+                    || (
+                        bar
+                    )
+            `
+        },
+        {
+            code: unIndent`
+                foo
+                                || (
+                                    bar
+                                )
+            `
+        },
+        {
+            code: unIndent`
                 var foo =
                         1;
             `,
@@ -7423,6 +7439,21 @@ ruleTester.run("indent", rule, {
                 ]
             `,
             errors: expectedErrors([3, 0, 4, "Punctuator"])
+        },
+        {
+            code: unIndent`
+                foo
+                        || (
+                                bar
+                            )
+            `,
+            output: unIndent`
+                foo
+                        || (
+                            bar
+                        )
+            `,
+            errors: expectedErrors([[3, 12, 16, "Identifier"], [4, 8, 12, "Punctuator"]])
         },
         {
             code: unIndent`

